src/core/ddsc/tests/psmx.c

```diff
@@ -1264,6 +1391,164 @@ CU_Test (ddsc_psmx, basic)
+/// - Create a configuration with a psmx interface capable of shared memory and don't specify a locator.
+/// - Create a domain using this configuration.
+/// - Query whether shared memory is supported.
+/// - Assert that there is exactly one psmx instance.
+/// - Assert that the psmx instance has a nonempty instance_name.
+/// - Create some entities
+/// - Check if shared memory is enabled.
+/// - Expectation: Shared memory is enabled iff the psmx interface supports it (use queried value).
+/// - Delete the domain
+/// - Check the function call counts of the dummy psmx.
+/// - Expectation: The counts match expectations. In particular, create counts must match their delete counterpart.
+/// 
+CU_Test(ddsc_psmxif, shared_memory)
+{
+  const char* CDDS_PSMX_NAME = NULL;
+  ddsrt_getenv("CDDS_PSMX_NAME", &CDDS_PSMX_NAME);
+  ddsrt_setenv("CDDS_PSMX_NAME", "dummy"); // Make `create_participant()` use the dummy psmx.
+  dummy_mockstats_t dmock, dmock_expected;
+  char strbuf1[512];
+  char strbuf2[512];
+  size_t strbuf_size = sizeof(strbuf1);
+  {
+    // Check that the data types I'm planning to use are actually suitable for use with shared memory.
+    dds_data_type_properties_t props;
+    props = dds_stream_data_types(SC_Model_desc.m_ops);
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+    props = dds_stream_data_types(PsmxType1_desc.m_ops);
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+  }
+
+  const dds_domainid_t domainId = 3;
+  bool supports_shared_memory = false; // Overwritten by `plugin's supported_features()`.
+  bool specify_locator[] = {true, false};
+  uint8_t locator_in[16];
+  memset(locator_in, 0x0, sizeof(locator_in)); // Avoid warning 'uninitialized value'.
+  ((uint64_t*)locator_in)[0] = (uint64_t)0x4a4d203df6996395;
+  ((uint64_t*)locator_in)[1] = (uint64_t)0xe1412fbecc2de4b6;
+
+  int N = sizeof(specify_locator) / sizeof(specify_locator[0]);
+  for (int i = 0; i < N; ++i)
+  {
+    const uint8_t* locator = specify_locator[i] ? locator_in : NULL;
+    dds_entity_t participant = create_participant(domainId, locator);
+    CU_ASSERT_FATAL(participant > 0);
+    dds_entity_t domain = dds_get_parent(participant);
+    CU_ASSERT_FATAL(domain > 0);
+    if ( specify_locator[i] ) {
+      // Check that I get the same locator that I provided with the config.
+      ddsi_locator_t locator_out;
+      CU_ASSERT_FATAL(domain_get_psmx_locator(domain, &locator_out));
+      CU_ASSERT_FATAL(memcmp(locator_in, locator_out.address, sizeof(locator_out.address)) == 0);
+    }
+
+    {
+      // Query whether shared memory is supported.
+      dds_domain* dom = NULL;
+      CU_ASSERT_FATAL(domain_pin(domain, &dom));
+      CU_ASSERT_FATAL(dom->psmx_instances.length == 1); // There shall be exactly one psmx instance.
+      struct dummy_psmx* dpsmx = (struct dummy_psmx*)dom->psmx_instances.instances[0];
+      dpsmx->mockstats_get_ownership(&dmock);
```

Comment:
I don't even understand this. Well, maybe I do, but it seems overly complicated for what you're trying to achieve.

This is just test code. We know there is exactly one such plugin loaded. We know the plugin. We can even *link against it from the test code* (these are just libraries after all).

So instead of carefully extracting a pointer to the `dds_psmx` struct, why not store it in a global variable inside the plugin and have the plugin export a function that returns the address of it? Then you just declare that function in `psmx_dummy_public.h` and call it.

Indeed, instead of doing that with the `dds_psmx` one, I would just let it store a global `dummy_mockstats_t` and return a pointer to that. No need to fiddle with the innards.

Reply:
The issue with using a global variable inside the plugin, is that it's gone as soon as you delete the domain, so you would miss all of the calls done during cleanup.

Reply:
`dlclose` doesn't unconditionally unload the library, it only unloads it when there is no further reference to it. `dlopen` is one way to get such a reference, linking the executable with the library is another.

So you could simply add the dummy plugin to the set of libraries to link the test executable with, or you could let the test also `dlopen`/`dlcose` it. (I'd probably go for the first option, but they are equivalent.)